src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,715 @@
+                            ++_index;
+                            return true;
+                        }
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override AppendPrependIterator<TSource> Append(TSource element)
+            {
+                if (RunningOnCreatingThread && _appended.Append(element, _appCount))
+                {
+                    return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount, _appended, _appCount + 1);
+                }
+
+                return new Append1Iterator<TSource>(this, element);
+            }
+
+            public override AppendPrependIterator<TSource> Prepend(TSource element)
+            {
+                if (RunningOnCreatingThread && _prepended.Prepend(element, _preCount))
+                {
+                    return new AppendPrependNIterator<TSource>(_source, _prepended, _preCount + 1, _appended, _appCount);
+                }
+
+                return new Prepend1Iterator<TSource>(this, element);
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + _preCount + _appCount;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count + _preCount + _appCount];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
+                }
+
+                Debug.Assert(index == _preCount);
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>() : new List<TSource>(count + _appCount + _preCount);
```

Comment:
I set it to be `Math.Max(_appCount + _preCount, 4)` in this case, 4 being `List<T>._defaultCapacity`. Similarly for the single-item versions I set it to `new List<TSource(4)` since there's always going to be at least one `Add()` and hence on move from 0 to 4.